### DISCUSSION TITLE
CTT Data Access PercentDeadband gets BadMonitoredItemfilterUnsupported

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1318] [General]
Hi,
 in our server we have AnalogItemType Values and CTT fails with most of the PercentDeadband Tests:
![image](https://github.com/user-attachments/assets/4db47550-a0c5-437f-8bde-b223d813d94c)
Is it not implemented yet ? How can we support here. 

Greetings Johann

**Comment by kevinherron** [comment]
Percent deadband filters are not implemented. You're welcome to have a go at it.

One of the things that is tricky about implementing this is that the EURange and EngineeringUnits are separate property nodes, which makes obtaining these a little more work. The easiest and most performant option would be to lock these in at the time the filter is created rather than trying to read them every time the filter evaluates, but if the properties are writable by Clients then this may not end up being viable, unless you also added logic to update the filter any time these changed.

**Comment by johannwesely** [comment]
Hello, 
we use some AnalogSignals which are Subtypes of AnalogItemType in our Server and we want to go into certification. But we still have the Percent Deadband Ctt issue using milo 1.0.1. My Question is do you plan to integrate this in near future or what are our options here?
 
Greetings Johann

**Reply by kevinherron** [reply]
No plan to implement this any time soon.

AFAIK percent deadband is not required by any of the UA server profiles. You don't need to implement it to get certified.

**Comment by johannwesely** [comment]
You Saved my Day. I thought i need it for certification. Can you tell me what DA Profile means in ctt Tool ?

**Reply by johannwesely** [reply]
because in the opcuafoundation application form for certification you have to enter all datatypes in "UA Data Types" occuring in your server and there is also AnalogItemType selectable.

**Reply by kevinherron** [reply]
The DA Profile in CTT is testing all the stuff defined in OPC UA Part 8 (Data Access).

I think just don't select the AnalogItemType, or any other DA stuff, even if it occurs in your server.

Might want to ask Alex or somebody else from OPC Foundation compliance team about how to fill out the application. It's been many years since I certified a product and it's no doubt a different process now.

**Comment by johannwesely** [comment]
So my Information right now i got from Sebastian Allmendinger last week is: "AnalogItems are not included in the 'Standard 2017 UA Server Profile'. However, once they are supported/available on the server, the 'Data Access Server Facet' will be included in the test. This will then include tests for AnalogItems among others." so this means i have to implement all the AnalogItem functionality at least. Can you point me to where i can start on this?   Thanks in Advance Johann

**Reply by kevinherron** [reply]
What's going to make this quite difficult is that in in order to implement percent deadband, at the time the Monitored Item is created you need to:
1. figure out of the target Node is even an AnalogItemType
2. if it is, read the EURange and hang onto it so the filter implementation can reference it

Creating a MonitoredItem and validating the filter is done in `SubscriptionManager::createMonitoredItem` and `SubscriptionManager::validateDataItemFilter`. 

The tricky thing is that when implementing functionality in the SDK you can't assume UaNode instances exist. To even figure out of a Node is a AnalogItemType you have to browse for its type definition and check if its AnalogItemType or a subclass. Then if it is, you have to browse for its property nodes and find the one with the "EURange" browse name. Then you have to read that Node to get the EURange value.

The actual filter logic is applied by `DataChangeMonitoringFilter.filter` called in `MonitoredDataItem::setValue`. Something about how filters are "installed" on a MonitoredItem will have to change so that you can pass along the EURange for the filter logic. Maybe instead of installing the `MonitoringFilter` UA structure in `BaseMonitoredItem::installFilter` an implementation of the filter can be installed instead or something.

All of this has to be considered in the code path that modifies Monitored Items as well.

All of this has to not impact the performance when creating or modifying Monitored Items that don't request a percent deadband filter.

**Comment by johannwesely** [comment]
Hello, 
i started implementation of percent Deadband filters u said that i cant assume UaNode instances exist but for me something like this works well: 
````
private static NodeId getTypeDefinition(OpcUaServer server, NodeId nodeId) throws UaException {
	
		// get Node from NodeManager
		UaNode node = server.getAddressSpaceManager().getManagedNode(nodeId).orElse(null);

		// get NodeId of TypeDefinition
		if (node instanceof UaVariableNode variableNode) {
			return variableNode.getTypeDefinitionNode().getNodeId();
		}

		return null;

	}
```` 
What is the reason i cant assume UaNode instances exist? Or how could a browse approach look like ? 

Greetings Johann

**Reply by kevinherron** [reply]
The reason the AddressSpace API looks the way it does is so that it's possible to implement an AddressSpace without actually having UaNode instances, which allows for large or dynamic address spaces that would otherwise be difficult to fit into memory.

I think right now the only functionality the SDK requires that you use UaNode for is Events. MonitoredItems do not currently require it, and probably shouldn't.
